### PR TITLE
Scorecard query templates

### PR DESCRIPTION
**Pull request by mollybsmith-noaa**
This pull request populates scorecard query templates depending on input from the user, and stores them in the Scorecard document.

### CODE REVIEW

**Review by randytpierce — APPROVED** [review]
looks fine. I like the template idea.

**Comment by mollybsmith-noaa** [comment]
I'm merging this now so Randy can proceed, but if Ian still wants to look at it I'd appreciate the feedback.

**Review by ian-noaa — COMMENTED** [review]
The templates look like a good idea! I wonder if there's some sort of JS templating engine (like Jinja in Python) we could use instead of needing to do our own template processing.

Otherwise, the biggest critique I have about the PR is that the `processScorecard()` function seems like it's in need of a refactor - having 6 levels of nested `forEach` calls is quite a lot and can cause us maintainability and performance issue down the road.

It's merged in and sounds like it's doing what we want but it might be good to refactor it sometime while it's still fresh in mind.

File: apps/scorecard/server/dataFunctions/processScorecard.js [hunk 1]
```diff
@@ -263,9 +325,65 @@ processScorecard = function (plotParams, plotFunction) {
...
+
+                // populate level in query template
+                if (localQueryTemplate.includes("{{level}}")) {
+                  localQueryTemplate = localQueryTemplate.replace(
+                    /\{\{level\}\}/g,
+                    levelText
+                  );
+                }
+
+                debugger;
```
[the remark this answers is not in the document]
Comment (mollybsmith-noaa) [reply apps/scorecard/server/dataFunctions/processScorecard.js]: Yeah, it was easier to just leave it in for the intermediate commits, but I made sure I got it at the end.

File: apps/scorecard/server/dataFunctions/processScorecard.js [hunk 2]
```diff
@@ -263,9 +325,65 @@ processScorecard = function (plotParams, plotFunction) {
...
+                  );
+                }
+
+                // populate forecastLength in query template
+                if (localQueryTemplate.includes("{{forecastLength}}")) {
+                  localQueryTemplate = localQueryTemplate.replace(
+                    /\{\{forecastLength\}\}/g,
+                    fcstlenText
+                  );
+                }
```
[the remark this answers is not in the document]
Comment (ian-noaa) [reply apps/scorecard/server/dataFunctions/processScorecard.js:446]: Hah. Of course, I landed on the easy example. In that case, the code could have a function for each special case. e.g.

```js
function populateForecastLength(/*required args*/) {
  // whatever needs to be done to populate the forecast length
}

function populateThreshold(/*required args*/) {
  // whatever needs to be done to populate the threshold values
}
```

It'd still improve the readability of the `processScorecard` function and would be a good entrypoint for unit testing the template replacements.

If things do get really complex, that could be where a templating engine like Mozilla's Nunjucks would come in.

[the remark this answers is not in the document]
Comment (mollybsmith-noaa) [reply apps/scorecard/server/dataFunctions/processScorecard.js:446]: All right, I can do this to simplify the code when the higher-priority functionality tasks are completed (this seems like medium priority?). Is that okay?

[the remark this answers is not in the document]
Comment (ian-noaa) [reply apps/scorecard/server/dataFunctions/processScorecard.js:446]: Definitely! It seems like a good thing to do after Bonny's deadline next week.

**Pull request merged**